src/libraries/Common/tests/TestUtilities/System/AssertExtensions.cs

```diff
@@ -412,24 +412,44 @@ public static void CollectionEqual<T>(IEnumerable<T> expected, IEnumerable<T> ac
 
             if (expectedCount != actualCount)
             {
-                throw new XunitException($"Expected count: {expectedCount}{Environment.NewLine}Actual count: {actualCount}");
+                Throw();
             }
 
             for (var i = 0; i < expectedCount; i++)
             {
                 T currentExpectedItem = expectedArray[i];
                 if (!actualItemCountMapping.TryGetValue(currentExpectedItem, out ItemCount countInfo))
                 {
-                    throw new XunitException($"Expected: {currentExpectedItem} but not found");
+                    Throw();
                 }
 
                 if (countInfo.Remain == 0)
                 {
-                    throw new XunitException($"Collections are not equal.{Environment.NewLine}Totally {countInfo.Original} {currentExpectedItem} in actual collection but expect more {currentExpectedItem}");
+                    Throw();
                 }
 
                 countInfo.Remain--;
             }
+
+            void Throw()
+            {
+                string exMessage = $"Collections are not equal.{Environment.NewLine}";
+                exMessage += $"Expected length: {expectedCount}, actual length: {actualCount}{Environment.NewLine}";
+
+                exMessage += $"Expected elements:{Environment.NewLine}";
+                foreach (T e in expected)
```

Comment:
Would it be better to skip items that were expected && present and only show the missing and extras? Collections might be huge in the general case with only one item wrong.

Reply:
This method currently stops at the first discrepancy. It will require more changes to make it resilient to missing elements and continue iterating. 
It can be done if we see it is really an issue to have very large collections logged.

Reply:
I agree with @danmoseley, the collections can be huge and in such cases, we would print too much information. I think that it's OK to print the actual items in case one is missing, but only if there are few of them (less than 10?).

Reply:
> It will require more changes to make it resilient to missing elements and continue iterating.

Just a few lines. Instead of throwing, keep appending to the StringBuilder. (Cut off at say 10 items so output is never huge). Then throw at the end with that message.

BTW possibly the method's comment should point out that it doesn't verify ordering. It's not useful for comparing two ordered Lists, for example.

Reply:
Done in https://github.com/dotnet/runtime/pull/68066/commits/5a0642884be1176e11a82da1da784309dfab5fed. I flipped the actualItemMap to expectedItemMap in order to log the actual results since those are more valuable than the expected ones.